### PR TITLE
prov/tcp: changing the recv to accommodate rma read/writes

### PR DESCRIPTION
**Pull request by vkrishna**
Signed-off-by: Venkata Krishna Nimmagadda <venkata.krishna.nimmagadda@intel.com>

### CODE REVIEW

File: prov/tcp/src/tcpx_progress.c [hunk 1]
```diff
@@ -117,18 +117,71 @@ static void process_rx_pe_entry(struct tcpx_pe_entry *pe_entry)
 	tcpx_pe_entry_release(pe_entry);
 }
 
-static void process_rx_queue(struct tcpx_ep *ep)
+static struct tcpx_pe_entry *
+tcpx_parse_hdr_and_get_entry(struct tcpx_rx_hdr *rx_hdr)
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: The function name implies that this should be 2 separate functions.

File: prov/tcp/src/tcpx.h [hunk 2]
```diff
@@ -184,9 +188,16 @@ enum tcpx_cm_state {
 	TCPX_EP_ERROR,
 };
 
+struct tcpx_rx_hdr {
+	struct ofi_op_hdr	hdr;
+	uint64_t		done_len;
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: 'done_len' should not be part of a header.  Is this really a wire structure? If not, I would rename 'hdr' to something else to avoid confusion.

File: prov/tcp/src/tcpx_comm.c [hunk 3]
```diff
@@ -57,42 +57,29 @@ int tcpx_send_msg(struct tcpx_pe_entry *pe_entry)
 	return FI_SUCCESS;
 }
 
-static int tcpx_recv_msg_hdr(struct tcpx_pe_entry *pe_entry)
+int tcpx_recv_field(SOCKET sock, void *buf, size_t buf_len,
+		    uint64_t *done_len, size_t start_offset)
```
Comment (shefty) [inline prov/tcp/src/tcpx_comm.c]: It looks odd seeing both done_len and start_offset.  Why doesn't done_len == start_offset?

File: prov/tcp/src/tcpx_cq.c [hunk 4]
```diff
@@ -94,6 +94,9 @@ void tcpx_pe_entry_release(struct tcpx_pe_entry *pe_entry)
 
 	tcpx_cq = container_of(cq, struct tcpx_cq, util_cq);
 
+	if (pe_entry->ep->cur_rx_entry == pe_entry)
+		pe_entry->ep->cur_rx_entry = NULL;
```
Comment (shefty) [inline prov/tcp/src/tcpx_cq.c]: This sort of change should be protected by some sort of lock.

File: prov/tcp/src/tcpx_progress.c [hunk 5]
```diff
@@ -117,18 +117,72 @@ static void process_rx_pe_entry(struct tcpx_pe_entry *pe_entry)
...
+					entry);
+
+		pe_entry->msg_hdr = rx_hdr->hdr;
+		pe_entry->msg_hdr.op_data = TCPX_OP_MSG_RECV;
+		pe_entry->done_len = sizeof(rx_hdr->hdr);
+
+		if (ofi_truncate_iov(pe_entry->msg_data.iov,
+				     &pe_entry->msg_data.iov_cnt,
+				     (ntohll(pe_entry->msg_hdr.size) -
+				      sizeof(pe_entry->msg_hdr))))
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: The purpose of calling this function isn't obvious to me.  Neither is error handling that occurs if it fails.

File: prov/tcp/src/tcpx.h [hunk 6]
```diff
@@ -213,18 +213,23 @@ struct tcpx_fabric {
...
+		struct fi_rma_iov	rma_iov[TCPX_IOV_LIMIT];
+		struct fi_rma_ioc	rma_ioc[TCPX_IOV_LIMIT];
 	};
+};
+
+struct tcpx_msg_data {
+	size_t		iov_cnt;
+	struct iovec		iov[TCPX_IOV_LIMIT+1];
 	uint8_t			inject[TCPX_MAX_INJECT_SZ];
 };
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: The reason for separating msg_data from rma_data isn't obvious.  Please add more details to the commit messages!  Note that RMA writes can still be sent using inject.

File: prov/tcp/src/tcpx_comm.c [hunk 7]
```diff
@@ -76,18 +76,78 @@ int tcpx_recv_field(SOCKET sock, void *buf, size_t buf_len,
...
+static int tcpx_process_write(struct tcpx_pe_entry *pe_entry)
+{
+	return -FI_ENODATA;
+}
+
+static int tcpx_validate_rma_data(struct tcpx_pe_entry *pe_entry)
+{
+	return -FI_ENODATA;
+}
+static int tcpx_process_rma_data(struct tcpx_pe_entry *pe_entry)
```
Comment (shefty) [inline prov/tcp/src/tcpx_comm.c]: I would add an 'rx' tag somewhere in these names, to indicate that this is the receive side handling of an rma.

File: prov/tcp/src/tcpx_comm.c [hunk 8]
```diff
@@ -76,18 +76,78 @@ int tcpx_recv_field(SOCKET sock, void *buf, size_t buf_len,
...
+				  sizeof(pe_entry->rma_data))) {
+		switch (pe_entry->msg_hdr.op) {
+		case ofi_op_read_req:
+		case ofi_op_read_rsp:
+		case ofi_op_write:
+			ret = tcpx_process_rma_data(pe_entry);
+			if (ret)
+				return ret;
+		}
+	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_comm.c]: This code seems misplaced.  The function as it currently stands reads only application data, which seems fairly clean.  (You could rename the call to reflect this).  What looks like is needed is a call to receive the tcpx headers.  I would isolate reading the headers into a single location.

File: prov/tcp/src/tcpx_progress.c [hunk 9]
```diff
@@ -146,14 +149,22 @@ static struct tcpx_pe_entry * tcpx_get_rx_pe_entry(struct tcpx_rx_hdr *rx_hdr)
...
-	case ofi_op_read_rsp:
 	case ofi_op_write:
-	case ofi_op_write_rsp:
-		/* todo  complete these cases for rma*/
+		pe_entry = tcpx_pe_entry_alloc(tcpx_cq);
+		if (!pe_entry)
+			return NULL;
+
+		pe_entry->msg_hdr = rx_hdr->hdr;
+		pe_entry->msg_hdr.op_data = TCPX_OP_MSG_RECV;
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: The op code looks wrong

File: prov/tcp/src/tcpx.h [hunk 10]
```diff
@@ -222,8 +222,8 @@ struct tcpx_rma_data {
...
 
 struct tcpx_msg_data {
-	size_t		iov_cnt;
-	struct iovec		iov[TCPX_IOV_LIMIT+1];
+	size_t			iov_cnt;
+	struct iovec		iov[TCPX_IOV_LIMIT+2];
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: We should try to keep all headers within a single iovec, to avoid needing to access multiple iov entries.

File: prov/tcp/src/tcpx_attr.c [hunk 11]
```diff
@@ -80,7 +80,9 @@ static struct fi_fabric_attr tcpx_fabric_attr = {
 };
 
 struct fi_info tcpx_info = {
-	.caps = FI_MSG | FI_SEND | FI_RECV,
+	.caps = FI_MSG | FI_SEND | FI_RECV |
+	FI_RMA | FI_WRITE | FI_REMOTE_READ |
```
Comment (shefty) [inline prov/tcp/src/tcpx_attr.c]: Commit message only calls out rma write.  Is setting FI_REMOTE_READ correct?  If not, it should be removed from this patch.  If it is, the commit message should be updated.  And please provide more details in the commit messages than a single line! :)

File: prov/tcp/src/tcpx_attr.c [hunk 12]
```diff
@@ -80,7 +80,9 @@ static struct fi_fabric_attr tcpx_fabric_attr = {
...
+	FI_REMOTE_WRITE,
```
Comment (shefty) [inline prov/tcp/src/tcpx_attr.c]: nit: indent caps continuation bits another line

File: prov/tcp/src/tcpx_comm.c [hunk 13]
```diff
@@ -88,7 +88,16 @@ static int tcpx_process_read_rsp(struct tcpx_pe_entry *pe_entry)
...
-	return -FI_ENODATA;
+	int i;
+
+	pe_entry->msg_data.iov_cnt = pe_entry->rma_data.rma_iov_cnt;
+	for ( i = 0 ; i < pe_entry->rma_data.rma_iov_cnt ; i++ ) {
+		pe_entry->msg_data.iov[i].iov_base =
+			(void *) pe_entry->rma_data.rma_iov[i].addr;
+		pe_entry->msg_data.iov[i].iov_len =
+			pe_entry->rma_data.rma_iov[i].len;
+	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_comm.c]: Is this iov copy really necessary?  We're copying from within the same pe_entry structure.

On this note, the name 'pe_entry' seems confusing.  Maybe we should just have tx/rx_entry or tx/rx_req or tx/rx_something...  PE stood for progress engine in the socket provider, but we really don't have a progress engine here, just structures to track outstanding transmits or incoming receives.

File: prov/tcp/src/tcpx_progress.c [hunk 14]
```diff
@@ -113,7 +113,9 @@ static void process_rx_pe_entry(struct tcpx_pe_entry *pe_entry)
 done:
 	tcpx_cq_report_completion(pe_entry->ep->util_ep.rx_cq,
 				  pe_entry, ret);
-	dlist_remove(&pe_entry->entry);
+	if (!((pe_entry->entry.next == NULL) &&
+	      (pe_entry->entry.prev == NULL)))
```
Comment (shefty) [inline prov/tcp/src/tcpx_progress.c]: Why is this check needed?  This change looks unrelated and should be called out in the commit message.  (Have I mentioned adding more details to the commit messages?)  :)

File: prov/tcp/src/tcpx_rma.c [hunk 15]
```diff
@@ -107,7 +107,71 @@ static ssize_t tcpx_rma_readv(struct fid_ep *ep, const struct iovec *iov, void *
...
+		return -FI_EAGAIN;
+
+	assert(msg->iov_count <= TCPX_IOV_LIMIT);
+
+	data_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+
+	assert(!(flags & FI_INJECT) || (data_len <= TCPX_MAX_INJECT_SZ));
+
+	send_entry->msg_hdr.version = OFI_CTRL_VERSION;
+	send_entry->msg_hdr.op = ofi_op_write;
```
Comment (shefty) [inline prov/tcp/src/tcpx_rma.c]: The rxm provider ended up adding multiple pools based on the operation type.  Each pool had pre-formatted headers, so that it wasn't necessary to format them on every data transfer call.   We should add something similar for tcpx in order to reduce software overhead.  A lot of the fields in send_entry can be preset.

File: prov/tcp/src/tcpx_rma.c [hunk 16]
```diff
@@ -107,7 +107,71 @@ static ssize_t tcpx_rma_readv(struct fid_ep *ep, const struct iovec *iov, void *
...
+
+	tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	tcpx_cq = container_of(tcpx_ep->util_ep.tx_cq, struct tcpx_cq,
+			       util_cq);
+
+	send_entry = tcpx_pe_entry_alloc(tcpx_cq);
+	if (!send_entry)
+		return -FI_EAGAIN;
+
+	assert(msg->iov_count <= TCPX_IOV_LIMIT);
```
Comment (shefty) [inline prov/tcp/src/tcpx_rma.c]: This assert is <=, but the rma_iov_count assert below uses <.  < seems more appropriate.

File: prov/tcp/src/tcpx_rma.c [hunk 17]
```diff
@@ -107,7 +107,71 @@ static ssize_t tcpx_rma_readv(struct fid_ep *ep, const struct iovec *iov, void *
...
+	send_entry->msg_data.iov[0].iov_base = (void *) &send_entry->msg_hdr;
+	send_entry->msg_data.iov[0].iov_len = sizeof(send_entry->msg_hdr);
+
+	assert(msg->rma_iov_count < TCPX_IOV_LIMIT);
+	memcpy(send_entry->rma_data.rma_iov, msg->rma_iov,
+	       msg->rma_iov_count * sizeof(msg->rma_iov[0]));
+	send_entry->rma_data.rma_iov_cnt = msg->rma_iov_count;
+
+	send_entry->msg_data.iov[1].iov_base = (void *) &send_entry->rma_data;
+	send_entry->msg_data.iov[1].iov_len = sizeof(send_entry->rma_data);
```
Comment (shefty) [inline prov/tcp/src/tcpx_rma.c]: rma_data and msg_hdr should be merged into a single iovec entry. Plus we're copying msg->rma_iov to send_entry->rma_data, then copying that to send_entry->msg_data.

File: prov/tcp/src/tcpx_rma.c [hunk 18]
```diff
@@ -107,7 +107,71 @@ static ssize_t tcpx_rma_readv(struct fid_ep *ep, const struct iovec *iov, void *
...
+
+	send_entry->msg_hdr.flags = htonl(send_entry->msg_hdr.flags);
+	send_entry->ep = tcpx_ep;
+	send_entry->context = msg->context;
+	send_entry->done_len = 0;
+
+	fastlock_acquire(&tcpx_ep->queue_lock);
+	dlist_insert_tail(&send_entry->entry, &tcpx_ep->tx_queue);
+	fastlock_release(&tcpx_ep->queue_lock);
+	return FI_SUCCESS;
```
Comment (shefty) [inline prov/tcp/src/tcpx_rma.c]: Does this attempt to progress the write, or is that only done later when reading the CQ?  If the latter, it would be best to attempt progress on the write here.

File: prov/tcp/src/tcpx_cq.c (hunk 4 above)
Comment (vkrishna) [reply prov/tcp/src/tcpx_cq.c]: Yes it is already protected by queue_lock from this calling function tcpx_progress()

**Comment by vkrishna** [comment]
Thanks @shefty for the feedback. I will update the commit messages appropriately.

File: prov/tcp/src/tcpx.h (hunk 6 above)
Comment (vkrishna) [reply prov/tcp/src/tcpx.h]: I want to have rma_data before I get the actual message data so that I can directly place it/read it from, at the remote side.  if i have a same structure holding both rma and msg iov, then it is not possible since i am using scatter-gather read call to get data. 

one way to optimize is to coalesce rma iov with msg hdr, thus decreasing number of iovs.

File: prov/tcp/src/tcpx_comm.c (hunk 13 above)
Comment (vkrishna) [reply prov/tcp/src/tcpx_comm.c]: This is on the receive side after getting the rma data, it is validated and copied to iov such that the readv can be called on socket directly to get the data into destination. I will add these detail in commit messages.

**Comment by shefty** [comment]
Thanks - this is looking good!
